trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/PathProgramCache.java

```diff
@@ -53,8 +53,8 @@ public class PathProgramCache<LETTER> {
 
 	public PathProgramCache(final ILogger logger) {
 		mLogger = logger;
-		mKnownPathPrograms = new HashMap<>();
-		mTraceHashes = new ArrayList<>();
+		mKnownPathPrograms = new ConcurrentHashMap<>();
+		mTraceHashes = Collections.synchronizedList(new ArrayList<>());
```

Comment:
This seems intended to make this class thread-safe, which it wasn't intended to be by the original author.

I don't have a particular objection to this change (though it might have a bit of a performance impact, it should not be critical).

But it leads me to wonder which other classes are out there with similar problems, that will only be detected by hard-to-reproduce concurrency bugs. Do you have somewhere a documentation of which mutable objects are shared between threads, and how thread safety is ensured for them?